### PR TITLE
Add detailed guide for building ros2_rust packages

### PR DESCRIPTION
**Pull request by nnmm**
Also extend contributing.md and condense the readme a bit. See [here](https://github.com/ros2-rust/ros2_rust/blob/doc_articles/README.md) for a rendered version.

### CODE REVIEW

**Review by jhdcs — CHANGES_REQUESTED** [review]
Only a partial review at the moment, I'll have to look at the rest later...

File: README.md [hunk 1]
```diff
@@ -11,94 +8,50 @@ Build status
...
-# Install the cargo-ament-build plugin
-cargo install cargo-ament-build
+In a nutshell, the steps to get started are:
 
-# In your workspace directory (ideally an empty one), run 
-mkdir src
+```shell
+mkdir -p workspace/src && cd workspace
 git clone https://github.com/ros2-rust/ros2_rust.git src/ros2_rust
+docker build -t ros2_rust_dev - < src/ros2_rust/Dockerfile
```
Comment (esteve) [inline README.md:34]: I don't agree with recommending using Docker for developing `ros2_rust`. I use it myself, but the preferred way should be more in line with the rest of ROS2, which does not require Docker at all. I'm ok with adding a section about how to use Docker with `ros2_rust`, but should be in addition to explaining a Docker-less workflow.

**Review by esteve — CHANGES_REQUESTED** [review]

File: README.md (hunk 1 above)
Comment (nnmm) [reply README.md:34]: Sure. Do you mean that the method in the README should be the non-Docker way, or only that I should explain the non-Docker way in `Building.md`? I assumed that most people looking at the README would be interested in a quick and non-intrusive way of trying it out, hence I used the Docker setup.

Comment (esteve) [reply README.md:34]: I think recommending Docker as the way to develop `ros2-rust` packages or the library itself is more intrusive, as it requires setting up a piece of software that is not strictly needed for ROS2 development. If someone has set up a ROS2 workspace already, they already have everything they need, and don't require Docker.

So, what I suggest is that we keep the Docker-less workflow, and then add an optional chapter about how to use Docker with `ros2-rust` for those of us who prefer to use it.

Comment (nnmm) [reply README.md:34]: @esteve Done.

File: docs/Building.md [hunk 2]
```diff
@@ -0,0 +1,214 @@
...
+
+Next, clone `ros2_rust` into it. The rest of this article will assume that you have cloned it into a subdirectory named `src` like this:
+
+```shell
+# Make sure to run this in the workspace directory
+mkdir src
+git clone https://github.com/ros2-rust/ros2_rust.git src/ros2_rust
+```
+
+*Note: Once `rclrs` is published on crates.io, it's not technically needed anymore to clone the `ros2_rust` repo, and this section should be adapted to reflect that.*
```
[the remark this answers is not in the document]
Comment (nnmm) [reply docs/Building.md:31]: Done.

File: README.md [hunk 3]
```diff
@@ -11,94 +8,50 @@ Build status
...
-
-```
-# Do this in a new terminal
+# In a new terminal (tmux window) inside Docker
 . ./install/setup.sh
 ros2 run rclrs_examples minimal_subscriber
 ```
 
-Enjoy!
+For an actual guide, see the following documents:
```
[the remark this answers is not in the document]
Comment (nnmm) [reply README.md:53]: This section has changed a bit, but I've kept the "more in-depth guide" wording.

File: README.md [hunk 4]
```diff
@@ -11,94 +8,50 @@ Build status
...
-- Clients and services
-- Tests
-- Documentation
+Lots of things are still missing however, see the [issue list](https://github.com/ros2-rust/ros2_rust/issues) for an overview.
 
-### Limitations
-
-- The `rclrs` interface is very limited for now and might not be idiomatic yet, any help and suggestion on the interface would be greatly appreciated
-- Due to the current ROS2 support of non-default clients, packages containing definitions of messages used in Rust crates must be present in the current workspace; otherwise message crates generation won't be triggered
+The client library is still rapidly evolving, and there are no stability guarantees.
```
[the remark this answers is not in the document]
Comment (nnmm) [reply README.md]: Done.

File: README.md [hunk 5]
```diff
@@ -11,94 +8,50 @@ Build status
...
-Here, the Foxy distribution of ROS 2 is used, but newer distributions can be used by simply replacing 'foxy' with the distribution name.
-
-```
-# First, make sure to have ROS 2 and vcstool installed (alternatively, install vcstool with pip):
-# sudo apt install ros-foxy-desktop ros-foxy-test-interface-files python3-vcstool libclang-dev clang
-# Install the colcon-cargo and colcon-ros-cargo plugins
-pip install git+https://github.com/colcon/colcon-cargo.git git+https://github.com/colcon/colcon-ros-cargo.git
-# Install the cargo-ament-build plugin
-cargo install cargo-ament-build
+In a nutshell, the steps to get started are:
```
[the remark this answers is not in the document]
Comment (nnmm) [reply README.md]: This should be solved now, the latest version of README.md doesn't require Docker.

File: docs/Building.md [hunk 6]
```diff
@@ -0,0 +1,230 @@
...
+
+This is a more detailed guide on how to build ROS 2 packages written in Rust that expands on the minimal steps in the README.
+
+In this guide, the Foxy distribution of ROS 2 is used, but newer distributions can be used by simply replacing 'foxy' with the distribution name everywhere.
+
+
+## Environment setup
+
+Building `rclrs` requires a standard [ROS 2 installation](https://docs.ros.org/en/foxy/Installation.html), and a few extensions.
+These extensions are: `colcon-cargo`, `colcon-ros-cargo`, `cargo-ament-build`. The former two are `colcon` plugins, and the latter is a `cargo` plugin.
```
Comment (jhdcs) [inline docs/Building.md:11]:
```suggestion
These extensions are: `colcon-cargo`, `colcon-ros-cargo`, `cargo-ament-build`. The first two are `colcon` plugins, and the third is a `cargo` plugin.
```
English is odd. Nothing's technically wrong with how you have it, but I think this looks a bit more "natural". Feel free to ignore, though.

File: docs/Building.md [hunk 7]
```diff
@@ -0,0 +1,230 @@
...
+sudo apt install -y git libclang-dev python3-pip python3-vcstool # libclang-dev is required by bindgen
+# Install these plugins for cargo and colcon:
+cargo install cargo-ament-build
+pip install git+https://github.com/colcon/colcon-cargo.git git+https://github.com/colcon/colcon-ros-cargo.git
+```
+
+
+### Choosing a workspace directory and cloning `ros2_rust`
+
+A "workspace directory", or just "workspace", is simply a directory of your choosing that contains your `ros2_rust` checkout and potentially other ROS 2 packages. It will also usually be your working directory for building. There is only one limitation: It must not contain the ROS 2 installation, so it can't be `/`, for instance. Note that this has nothing to do with a [`cargo` workspace](https://doc.rust-lang.org/book/ch14-03-cargo-workspaces.html).
```
Comment (jhdcs) [inline docs/Building.md:35]:
```suggestion
A "workspace directory", or just "workspace", is simply a directory of your choosing that contains your `ros2_rust` checkout and potentially other ROS 2 packages. It will also usually be your working directory for building. There is only one limitation: It must not contain the ROS 2 installation, so it can't be `/`, for instance. Note that this has **nothing** to do with a [`cargo` workspace](https://doc.rust-lang.org/book/ch14-03-cargo-workspaces.html).
```
A little extra emphasis to attract eyes to this sentence. Hopefully that will make it more likely that users/readers won't miss this and get confused.

File: docs/Building.md [hunk 8]
```diff
@@ -0,0 +1,230 @@
...
+
+```shell
+# Make sure to run this in the workspace directory
+mkdir src
+git clone https://github.com/ros2-rust/ros2_rust.git src/ros2_rust
+```
+
+*Note: Once `rclrs` is published on crates.io, it's not technically needed anymore to clone the `ros2_rust` repo, and this section will be modified to reflect that.*
+
+### Using the Docker image
```
Comment (jhdcs) [inline docs/Building.md:49]:
```suggestion
### ALTERNATIVE: Using the Docker image
```
This suggestion may not be needed.

File: docs/Building.md [hunk 9]
```diff
@@ -0,0 +1,230 @@
...
+
+
+## Troubleshooting
+
+If something goes very wrong and you want to start fresh, make sure to delete all `install*`, `build*`, `target`, and `.cargo` directories. Possibly update the Docker container if it is out of date.
+
+
+### Package identification
+
+When you forgot to source the ROS 2 installation, you'll get this error:
```
Comment (jhdcs) [inline docs/Building.md:217]:
```suggestion
If you forget to source the ROS 2 installation, you'll get this error:
```

**Review by jhdcs — CHANGES_REQUESTED** [review]
Just a few more suggested changes. That being said, I like how this is turning out!

File: docs/Building.md (hunk 9 above)
Comment (nnmm) [reply docs/Building.md:217]: Thanks, good catch!

File: docs/Building.md (hunk 8 above)
Comment (nnmm) [reply docs/Building.md]: You're right that it wasn't clear that this is not always required. I restructured it so that the "manual installation" and "docker" sections are next to each other, and marked as alternatives.

File: docs/Building.md [hunk 10]
```diff
@@ -0,0 +1,233 @@
...
+If you forget to source the ROS 2 installation, you'll get this error:
+
+> ERROR:colcon.colcon_core.package_identification:Exception in package identification extension 'python_setup_py'
+
+Source the ROS 2 installation and try again.
+
+
+### Failed to resolve patches
+
+When you've deleted your `install` directory but not your `.cargo` directory, you'll get this error:
```
Comment (jhdcs) [inline docs/Building.md:229]:
```suggestion
If you've deleted your `install` directory but not your `.cargo` directory, you'll get this error:
```

**Review by jhdcs — CHANGES_REQUESTED** [review]
I apologize for not catching this one earlier...

File: docs/Building.md (hunk 10 above)
Comment (nnmm) [reply docs/Building.md]: Done!

**Review by jhdcs — APPROVED** [review]

**Comment by nnmm** [comment]
Thanks @jhdcs! Any more changes requested from your side, @esteve?

**Review by jhdcs — APPROVED** [review]

**Review by esteve — CHANGES_REQUESTED** [review]

File: docs/Contributing.md [hunk 11]
```diff
@@ -1,4 +1,4 @@
-# Contributing
+# Contributing to `ros2_rust`
 Contributions to `ros2_rust` are very welcome!
 
 Work is coordinated in the [Matrix chat](https://matrix.to/#/+rosorg-rust:matrix.org). It's a good idea to check in with the other contributors there before starting development on a new feature.
```
Comment (esteve) [inline docs/Contributing.md:4]: While Matrix is nice, I personally prefer if the focus is on GitHub. It's hard to keep track of conversations on Matrix, and there's no trail of who is working on what. So I'd advise contributors to first check GitHub, file an issue for what they want to work on, and then perhaps talk about it on Matrix, but again, technical discussions on a chat are hard to follow.

File: docs/Building.md [hunk 12]
```diff
@@ -0,0 +1,230 @@
...
+
+<!--- These steps should be kept in sync with README.md --->
+```shell
+# Install Rust, e.g. as described in https://rustup.rs/
+# Install ROS 2 as described in https://docs.ros.org/en/foxy/Installation.html
+# Assuming you installed the minimal version of ROS 2, you need these additional packages:
+sudo apt install -y git libclang-dev python3-pip python3-vcstool # libclang-dev is required by bindgen
+# Install these plugins for cargo and colcon:
+cargo install cargo-ament-build
+pip install git+https://github.com/colcon/colcon-cargo.git git+https://github.com/colcon/colcon-ros-cargo.git
```
[the remark this answers is not in the document]
Comment (nnmm) [reply docs/Building.md]: Sure, done

**Review by esteve — CHANGES_REQUESTED** [review]

File: docs/Contributing.md (hunk 11 above)
Comment (nnmm) [reply docs/Contributing.md]: Ok, I rephrased this.

**Comment by esteve** [comment]
@nnmm the `Contributing.md` file should go back to the root and renamed back to `CONTRIBUTING.md`. GitHub uses that convention so that contributors get a notification before they submit code (see https://github.blog/2012-09-17-contributing-guidelines/). This is in accordance with other projects not specific to ROS2 as well (e.g. https://github.com/django/django/blob/main/CONTRIBUTING.rst, https://github.com/pocoproject/poco/blob/master/CONTRIBUTING.md)

**Comment by nnmm** [comment]
> the Contributing.md file should go back to the root and renamed back to CONTRIBUTING.md. GitHub uses that convention so that contributors get a notification before they submit code (see https://github.blog/2012-09-17-contributing-guidelines/). This is in accordance with other projects not specific to ROS2 as well (e.g. https://github.com/django/django/blob/main/CONTRIBUTING.rst, https://github.com/pocoproject/poco/blob/master/CONTRIBUTING.md)

@esteve if you look up a newer article, GitHub also checks the `docs` subdirectory. I'd also be surprised if it needs to be uppercase, but I renamed it just to make sure.

**Comment by esteve** [comment]
@nnmm you're right about the `docs/` folder (https://docs.github.com/en/communities/setting-up-your-project-for-healthy-contributions/setting-guidelines-for-repository-contributors), but in that article seems to always refer to the file as `CONTRIBUTING(.md)`, so perhaps it needs to be in uppercase anyway.

**Comment by nnmm** [comment]
Yep, it's uppercase now.

**Comment by esteve** [comment]
@nnmm that was all, thanks for addressing my feedback :slightly_smiling_face:

**Review by esteve — APPROVED** [review]
